torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -487,9 +504,12 @@ void ProcessGroupAgent::listenLoop() {
 }
 
 void ProcessGroupAgent::pollTimedOutRPCs() {
-  while (!shutdown_.load()) {
-    std::chrono::milliseconds sleepTime;
+  while (true) {
     std::unique_lock<std::mutex> lock{futureMutex_};
+    if (!rpcRunning_.load()) {
```

Comment:
We now hold the lock from when we check `rpcRunning` to until we sleep, and the lock is also held in `shutdown()` before we set `rpcRunning`.